lib/Monads/WhileLoopRules.thy

```diff
@@ -383,6 +383,22 @@ lemma whileLoop_wp:
   \<lbrace> I r \<rbrace> whileLoop C B r \<lbrace> Q \<rbrace>"
   by (rule valid_whileLoop)
 
+lemma whileLoop_wp':
+  "(\<And>r. \<lbrace> \<lambda>s. I r s \<and> C r s \<rbrace> B r \<lbrace> I \<rbrace>) \<Longrightarrow> \<lbrace> I r \<rbrace> whileLoop C B r \<lbrace> I \<rbrace>"
```

Comment:
The name `_wp'` doesn't really work, because the rule forces the post condition to be the invariant, so it's not a `wp` rule any more. Maybe `whileLoop_inv`? (not sure if that one is taken already)

Reply:
Hmm, it's not taken but there's a bit of a conceptual nameclash with the `whileLoop_inv` definition and 
```
lemma whileLoop_wp_inv [wp]:
  "⟦ ⋀r. ⦃λs. I r s ∧ C r s⦄ B r ⦃I⦄; ⋀r s. ⟦I r s; ¬ C r s⟧ ⟹ Q r s ⟧
   ⟹ ⦃ I r ⦄ whileLoop_inv C B r I M ⦃ Q ⦄"
```

Reply:
That does actually mesh, I think. The `wp_inv` variant is a wp rule, and the `inv` variant would not be, but is related. But the definition name would be confusing. How about `whileLoop_hoare_inv`  or `whileLoop_valid_inv`?